### PR TITLE
[Feature][Spec Decode] Reuse the common proposer for auxiliary draft caches

### PR DESCRIPTION
**Pull request by lijiahang226**
### What this PR does / why we need it?

Refs #15665

Reuse the existing Eagle proposer for draft models with auxiliary attention caches. Initialize draft metadata builders from the runner's resolved attention groups, backends and cache specs, keep each draft step's buffers independent, and advance positions once per step from the accepted endpoint.

This is the common proposer part of GLM-5.3-Flash MTP support. Full-model validation also includes #16214 and the KeyPool integration in #16253 (with its dependencies). Prerequisite commits are excluded from this branch.

### Does this PR introduce _any_ user-facing change?

Yes. Models with auxiliary draft caches can use the existing MTP proposer path with consistent metadata across speculative steps. Cache allocation and grouping retain the existing contract.

### How was this patch tested?

- Added `tests/ut/spec_decode/test_auxiliary_cache_draft_metadata.py` for attention-group reuse, independent draft buffers, position advancement and accepted-token rollback.
- Previous integration validation at `733d5a65a61a34faaede500c97fffcd3394c5ca2`: 139 related CPU tests passed, 13 skipped; two CPU-input/NPU-only operator failures also reproduced on the baseline. Five additional checks with real metadata builders passed.
- A3, GLM-5.3-Flash-w8a8, TP8 with expert parallelism, MTP=3, target FULL_DECODE_ONLY: eight smoke cases passed and a single 3500-input/1500-output request completed.
- That integration used #16214 at `959a13d`; its newer runner revision still needs graph-mode revalidation. GitHub pre-commit and CPU UT passed on this branch (4180 passed, 38 skipped); device CI awaits a maintainer test label.

- vLLM main: https://github.com/vllm-project/vllm/commit/b2f685834a6456197e7033966fdef52a23f1abcd

### CODE REVIEW

File: vllm_ascend/spec_decode/llm_base_proposer.py [hunk 1]
```diff
@@ -447,6 +447,62 @@ def load_model(self, model: nn.Module) -> None:
...
+                if names:
+                    group = copy.copy(runner_group)
+                    group.layer_names = names
+                    group.create_metadata_builders(
+                        self.vllm_config,
+                        self.device,
+                        # Preserve storage pages, even when the block table is
+                        # split into smaller kernel blocks.
+                        kernel_block_size=None,
+                        num_metadata_builders=self.num_speculative_tokens if set(names) <= auxiliary_layers else 1,
```
Comment (slippersss) [inline vllm_ascend/spec_decode/llm_base_proposer.py:474]: `num_metadata_builders` may not mean to be used in this way. Here is tricky, so please add comment to explain your usage.

File: vllm_ascend/spec_decode/llm_base_proposer.py [hunk 2]
```diff
@@ -447,6 +447,62 @@ def load_model(self, model: nn.Module) -> None:
                 else self.model.mask_hidden.view(self.hidden_size)
             )
 
+    def initialize_attn_backend(self, kv_cache_config, kernel_block_sizes=None):
+        layers = get_layers_from_vllm_config(self.vllm_config, AttentionLayerBase)
+        auxiliary_layers = {
+            name for name in self._draft_attn_layer_names if getattr(layers[name], "cache_role", "kv") != "kv"
+        }
+        if not auxiliary_layers:
+            return super().initialize_attn_backend(kv_cache_config, kernel_block_sizes)
```
Comment (slippersss) [inline vllm_ascend/spec_decode/llm_base_proposer.py:456]: Please double check and make sure that only glm53flash walks into this method.

File: vllm_ascend/spec_decode/llm_base_proposer.py [hunk 3]
```diff
@@ -447,6 +447,62 @@ def load_model(self, model: nn.Module) -> None:
...
+        slots = torch.full_like(positions, PADDING_SLOT_ID)
+        if common.num_reqs and table.shape[1]:
+            rows = torch.arange(positions.numel(), device=positions.device)
+            request = torch.searchsorted(common.query_start_loc[1 : common.num_reqs + 1], rows, right=True)
+            page = positions // block_table.block_size
+            valid = (request < common.num_reqs) & (positions >= 0) & (page < table.shape[1])
+            valid &= common.slot_mapping[: positions.numel()] >= 0
+            physical = table[request.clamp(max=common.num_reqs - 1), page.clamp(0, table.shape[1] - 1)].long()
+            valid &= physical >= 0
+            slots = torch.where(valid, physical * block_table.block_size + positions % block_table.block_size, slots)
```
Comment (slippersss) [inline vllm_ascend/spec_decode/llm_base_proposer.py:501]: Maybe dcp is not compatible with this? If yes and it is not adapted now, please raise error if using dcp.

File: vllm_ascend/spec_decode/llm_base_proposer.py [hunk 4]
```diff
@@ -1800,11 +1857,29 @@ def attn_update_stack_num_spec_norm(
...
+        if draft_index == 1 and len(self.draft_attn_groups) > 1 and not self.uses_mrope:
+            # Padded verification includes rejected tokens. All cache groups
+            # must resume drafting from the selected accepted endpoint.
+            common_attn_metadata.seq_lens = common_attn_metadata.seq_lens.clone()
+            common_attn_metadata.seq_lens[:batch_size] = used_update_positions[:batch_size] + 1
+            common_attn_metadata.seq_lens_cpu = None
+            common_attn_metadata._seq_lens_cpu = None
+            common_attn_metadata.num_computed_tokens_cpu = None
+            common_attn_metadata._num_computed_tokens_cpu = None
+            common_attn_metadata._num_computed_tokens_cache = None
```
Comment (slippersss) [inline vllm_ascend/spec_decode/llm_base_proposer.py:1881]: 1. Weird condition to enter here for glm53flash. Can we find a better way? Something more explicit.
2. Maybe `num_rejected_tokens_gpu` should be used instead of borrowing `used_update_positions`？

**Review by slippersss — CHANGES_REQUESTED** [review]

**Comment by slippersss** [comment]
BTW, e2e and doc should be added also.

**Comment by lijiahang226** [comment]
Closing this one deliberately as part of cleaning up my open PR backlog. Same situation as the sibling spec-decode PR: merge conflicts plus a missing maintainer test label, untouched for ~10 days, and I no longer plan to land it in this form.

The head commit `bce6789c7` stays on my fork branch `glm53-v5-mtp` if this needs to be revived later.
